blocks/bricks/attention.py

```diff
@@ -8,15 +8,140 @@
 of the agent* or simply *states*.
 
 """
+from abc import ABCMeta, abstractmethod
+
 from theano import tensor
+from six import add_metaclass
 
-from blocks.bricks import (MLP, Identity, Initializable, Sequence,
+from blocks.bricks import (MLP, Identity, Brick, Initializable, Sequence,
                            Feedforward, Tanh)
 from blocks.bricks.base import lazy, application
-from blocks.bricks.parallel import Parallel
+from blocks.bricks.parallel import Parallel, Distribute
+from blocks.bricks.recurrent import recurrent, BaseRecurrent
+from blocks.utils import dict_union, dict_subset
+
+
+@add_metaclass(ABCMeta)
+class AbstractAttention(Brick):
+    """The common interface for attention bricks.
+
+    A generic attention mechanism functions as follows. Its inputs are the
+    state of the agent and the object of attention, which we call here
```

Comment:
So, this might just be me because I'm not familiar enough with the usual vocabulary, but at first sight these docstrings are still really hard to read for me. Agent? Object of attention? Attended? Composite state of the agent? For me, it's unclear from the docstring what any of these things are.

I guess there's a variety of options:

1. Assume less familiarity with these concepts, and explain more explicitly what things like "agent" and "object of attention" mean. Use examples, not abstractions! My suggestion of a beginning would be something like:

> An attention mechanism decides what part of the input to pay attention to at a given point in time. It does so by taking a particular part of input (e.g. a word in a sentence, or a patch of larger image) and the "state" of the neural network at that point in time, assigning each partial input a score of "importance".

> For example, in a neural translation model [ref] the "state" is the hidden state of the RNN that outputs our translation, representing our translation so far. The partial inputs in this case are the words in the source sentence. Based on what the network has translated so far the attention mechanism now decides how much attention the network should pay to each word in the source sentence.

> The partial inputs and the scores given to each are combined into a "glimpse". As such, a glimpse is a single representation of what is important to the neural network (the agent) in its current state.

> A simple example of a glimpse would be taking the weighted sum of all the partial inputs, weighted by their importance.

2. Intersperse with pseudocode/docstrings and/or graphviz diagrams; maybe that is easier than textual explanation
3. Leave it, and eventually make a tutorial which introduces these concepts in a more gentle manner